### PR TITLE
Pham hoang long interview

### PR DESCRIPTION
**Pull request by longcuxit**
### Requirement

- Save data to local storage: done (using `indexDB`)
- The persistent feature: done
- The edit feature: done
- The active/complete todo feature: done
- Unit test: 100% (ignore : `src/index.tsx`, `src/service/\*`)
- Typescript: 99%

### More feature

- Change layout to support mobile
- Add search box to filter keyword
- Show confirm dialog when remove item
- Change: delete all todo => delete completed todo
- Add send button when create new todo(support for mobile)
- Add loading UI when call api (add, update, remove) (but using local api so it's not easy to see it)
- Login page (incomplete)

### System

- Update lasted packages (if project not working, please remove `node_modules`, use node v14 and install again)
- Change store from reducer to sweet format `src/utils/Store` (inspired by `react-sweet-context`)
- Change code base structure from single page to multi pages `src/pages`

- `PartialCheckbox` component: Show checkbox in partial
- `Loader` component: Manager multiple loading state by push promise
- `AsyncModal` component: Manage multiple modals by push and pop

### CODE REVIEW

**Comment by hoanglong-pyco** [comment]
### update fences structure

  - `components`: can not import: `models`, `pages`, `service`, `store`
  - `models`: can not import: `components`, `pages`, `service`, `store`
  - `pages`: can not import from another page the same level, (only accept `pages/path`: using switch to another pages)
  - `service`: can not import `components`, `pages`, `store`
  - `store`: can not import `components`, `pages`
  - `utils`: can not import from another folder the same level

File: src/__test__/App.spec.tsx [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+const mockConnect = Service.connect as jest.Mock;
+
+describe("App", () => {
+  it("should connected service & match snapshot", async () => {
+    mockConnect.mockResolvedValueOnce(true);
+    const { container } = render(<App />);
+    await act(() => Promise.resolve());
+
+    expect(container.querySelector(".pages")).toBeInTheDocument();
```
Comment (phi0411141) [inline src/__test__/App.spec.tsx]: We'd suggest:
- Use `findBy*` API from testing-library to find a not-exist-yet element instead of `act`.
- It is recommended not to use the `querySelector` from DOM API, in this case, we think `getByText` is better.

**Comment by phi0411141** [comment]
Hi sir,
I am Phi, Manabie's frontend engineer, hearing that you would like get our detail feedback, so I will comment directly in the PR, hope it is okay to you.

File: src/__test__/App.spec.tsx [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    expect(container).toMatchSnapshot();
```
Comment (phi0411141) [inline src/__test__/App.spec.tsx]: Because you are mocking almost everything, so we think snapshot is not needed anymore. We suggest to use snapshot to catch real UI instead of the mocked one.

File: src/components/.eslintrc [hunk 3]
```diff
@@ -0,0 +1,21 @@
+{
+  "extends": "../../.eslintrc",
+  "rules": {
+    "no-restricted-imports": [
```
Comment (phi0411141) [inline src/components/.eslintrc]: Really appreciate that you are a few people that take care the import direction. At Manabie, we use this library instead https://github.com/sverweij/dependency-cruiser
- Centralize import restrict rules
- Generate violation reports
- Generate deps import image for visualization.

File: src/components/AsyncModal/AsyncModal.tsx [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+interface State {
+  modals: ModalStack[];
+  current?: ModalStack;
+}
+
+const initialState: State = { modals: Array<ModalStack>() };
+
+let idIncrement = 0;
+
+const store = new Store(initialState, ({ set, get, partial }) => {
```
Comment (phi0411141) [inline src/components/AsyncModal/AsyncModal.tsx]: Interesting approach for modal, excellent understanding for promise, but we suggest adding some comments because the logic here is hard to get, especially the `promise.finally` block

File: src/components/Loader/Wrapper.tsx [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+  useState,
+  useEffect,
+  ComponentType,
+  FunctionComponent,
+} from "react";
+import classNames from "classnames";
+import { Spinner } from "react-bootstrap";
+
+import { LoaderContainer, useLoading } from "./Store";
+import { LoaderWrapperProps } from "./type";
```
Comment (phi0411141) [inline src/components/Loader/Wrapper.tsx]: This is just personal taste but we found that putting the component props interface in same file does improve DX, engineers do not need to jump back and forth to see the props list

File: src/components/Loader/Wrapper.tsx [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+  const [loading] = useLoading();
+  const [isIn, setIsIn] = useState(false);
+
+  props.className = classNames(
+    props.className,
+    "Loader",
+    loading && "Loader-loading",
+    isIn && "in"
+  );
+  props.style = { ...props.style, "--loader-duration": duration + "ms" } as any;
```
Comment (phi0411141) [inline src/components/Loader/Wrapper.tsx]: I think we can just create new variables instead of mutating the `props` object. It could be fine in this case because you are just reassigning, but it may create bad habit for team's engineer, people will just `props.styles[--loader-duration] =xyz`, which is not correct anymore

File: src/components/Loader/Wrapper.tsx [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+  useEffect(() => {
+    if (loading && !isIn) requestAnimationFrame(() => setIsIn(true));
+    else if (!loading && isIn) {
+      const timeOut = setTimeout(() => setIsIn(false), duration);
+      return () => clearTimeout(timeOut);
+    }
+  }, [loading, isIn, duration]);
+
+  return createElement(
```
Comment (phi0411141) [inline src/components/Loader/Wrapper.tsx]: We wonder why you abuse the `createElement` API too much, we don't think it is friendly for newbie engineer in the team, and we think all of below cases can be replaced by JSX

File: src/pages/Login/Form.tsx [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+
+export default function FormLogin() {
+  const actions = useAuthAction();
+  const pushLoader = usePushLoader();
+
+  return (
+    <Form
+      onSubmit={(e) => {
+        e.preventDefault();
+        pushLoader(actions.login("username", "password"));
```
Comment (phi0411141) [inline src/pages/Login/Form.tsx]: We wonder why you submit a hardcode `username` and `password` here

File: src/pages/Todos/Form.tsx [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+  onSubmit,
+  onBlur,
+}: TodoFormProps) {
+  const [value, setValue] = useState(defaultValue);
+  const [loading, pushLoader] = useLoading();
+
+  const handleSubmit = async (e: FormEvent<HTMLFormElement>) => {
+    e.preventDefault();
+    if (!value) return;
+    await pushLoader(onSubmit(value));
```
Comment (phi0411141) [inline src/pages/Todos/Form.tsx]: maybe we need to trim right before submit?

File: src/pages/Todos/Item.tsx [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+
+import { TodoStatus } from "models/Todo";
+import { useTodoActions } from "store/Todos";
+import { useLoading, withLoader } from "components/Loader";
+import { useAsyncConfirm } from "components/AsyncModal";
+import { createLocalStorage } from "utils";
+import TodoForm from "./Form";
+import { TodoItemProps } from "./type";
+
+const askMeDelete = createLocalStorage("DeleteAnItem", true);
```
Comment (phi0411141) [inline src/pages/Todos/Item.tsx]: Nice UX for remember the asking for delete

File: src/pages/index.tsx [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+  useEffect(() => {
+    if (!auth) return;
+
+    const authToken = "Bear " + auth.accessToken;
+
+    const id = client.interceptors.request.use((request) => {
+      request.headers.Authorization = authToken;
+      return request;
+    });
+    return () => client.interceptors.request.eject(id);
```
Comment (phi0411141) [inline src/pages/index.tsx]: We wonder what if we need to call API outside of React context, we need to set the interceptor again?

File: src/service/api-frontend.ts [hunk 12]
```diff
@@ -1,29 +1,94 @@
...
+      if (!db.objectStoreNames.contains("toDoList")) {
+        db.createObjectStore("toDoList", {
+          keyPath: "id",
+          autoIncrement: true,
+        });
+      }
+    };
+
+    const db = await toAsync(request);
+    todoStore = createStoreSelect(db, "toDoList");
```
Comment (phi0411141) [inline src/service/api-frontend.ts:23]: We think it may be easier to extend and test if we do `dependency injection` for the `todoStore`

File: src/service/api-frontend.ts [hunk 13]
```diff
@@ -1,29 +1,94 @@
...
-    async createTodo(content: string): Promise<Todo> {
-        return Promise.resolve({
-            content: content,
-            created_date: new Date().toISOString(),
-            status: TodoStatus.ACTIVE,
-            id: shortid(),
-            user_id: "firstUser",
-        } as Todo);
-    }
+  async connect() {
```
Comment (phi0411141) [inline src/service/api-frontend.ts:9]: Nice indexDB usage, first PR that use indexDB.
But beware that indexDB cant be used in cognito mode for some browsers in 2021

File: src/utils/Store/Notifier.ts [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+  addListen(listener: Listener): () => void {
+    const { _listens } = this;
+    _listens.push(listener);
+    return () => {
+      _listens.splice(_listens.indexOf(listener), 1);
+    };
+  }
+
+  protected notify(): void {
+    this._listens.slice().forEach((listener) => listener());
```
Comment (phi0411141) [inline src/utils/Store/Notifier.ts:34]: We wonder why you need to `slice` here?

File: src/utils/Store/Store.ts [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+    Value = State,
+    Flags extends any[] = never
+  >(
+    selector?: HookSelector<State, Value, Flags>
+  ): UseHook<Flags, Value, Action> {
+    const select = (selector || ((v) => v)) as Select;
+
+    return (...flags: Flags): [Value, Action] => {
+      const subscriber = useContext(this.Context);
+      const [state, setState] = useState(() =>
```
Comment (phi0411141) [inline src/utils/Store/Store.ts]: You store the `store` value inside another `useState`, which can be suspected to be staled in react concurrent and we cannot get the single source of truth for the `store` value anymore.
Recommend to use `useSyncExternalStore` instead.

File: src/utils/Store/Store.ts [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+
+    return Subscriber;
+  }
+
+  createHook<
+    Select extends HookSelect<State, Flags, Value>,
+    Value = State,
+    Flags extends any[] = never
+  >(
+    selector?: HookSelector<State, Value, Flags>
```
Comment (phi0411141) [inline src/utils/Store/Store.ts]: just a small suggestion, if you would like to create your own library, maybe you need allow user to pass an `equality function`

File: src/utils/__test__/index.spec.tsx [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+    storage.set(true);
+    expect(storage.get()).toBe(true);
+  });
+});
+
+describe("utils/Promise.delay:", () => {
+  it("initial default value", async () => {
+    const now = Date.now();
+    await Promise.delay();
+    expect(Date.now() - now >= 0).toBe(true);
```
Comment (phi0411141) [inline src/utils/__test__/index.spec.tsx]: Actually I think this expect statement cannot tell anything, because the `delay` is a promise, you await it mean it will be pushed to the queue, thus of course `Date.now() - now` is `true` anyway

File: src/utils/polyfill.ts [hunk 18]
```diff
@@ -0,0 +1,13 @@
...
+  interface ObjectConstructor {
+    keys<O>(object: O): (keyof O)[];
+  }
+
+  interface PromiseConstructor {
+    delay(time: number): Promise<void>;
+  }
+}
+
+Promise.delay = (time) => new Promise((next) => setTimeout(next, time));
```
Comment (phi0411141) [inline src/utils/polyfill.ts]: I think it is better to create a `delay` function instead of modify the `Promise` object, it may cause misunderstanding for other people (because they cant search in MDN)

**Comment by phi0411141** [comment]
We really appreciate that you spend time for our challenge (4k6 additions is huge). Big thanks.
Overall:
**Pros:**
  - A very good pull request, you have shown your skills.
  - Create a state management library is a great achievement.
  - Nice usage of promise for dialog and state control.
  - Nice UI re-decoration.
  - Handle a lot of edge cases, plus confirmation on deletion.
  - Have a lot of tests that cover most of the usage.
  - Have a proper project structure + good import direction restriction.
  - High level understanding of typescript, which is very rare to see in vietnam.
**Cons:**
  - Still use a lot of DOM API (querySelector) in test, let's switch to `getByText`, `getByRole`, that is something important to user (accessibility), user can't use the `querySelector` :v: 
  -  You are mixing state control in-and-out of React, which make most of your tests need `act`, it is not friendly in my opinion
  - This PR is extremely complex, my teammate give up reading it in a glance. You are doing low-level API (state management, dialog control, async confirmation), it is better if you can add comment describe why you are doing like that. We can't find a team of full senior members with good technical knowledge :(
  - Your state management library is coupled with React and it's usage is more complex than it should be. Maybe store APIs like https://github.com/pmndrs/zustand is simpler in my opinion.
  - Advance usage of promise may hurt the developer experience, as human mental thinking is very limit, if your code flow jumps from here to there too much, it creates confusion. But anyway, It's just trade-offs for more convenience.

File: src/utils/polyfill.ts (hunk 18 above)
Comment (hoanglong-pyco) [reply src/utils/polyfill.ts]: Ok! i think so, it not good
I created a new function for this function but not in this PR, because it's a lot of code, I can't pay attention to all
Thank for this.

File: src/utils/Store/Store.ts (hunk 16 above)
Comment (hoanglong-pyco) [reply src/utils/Store/Store.ts:90]: :) I just want to show my power.
so  i think public this library, be cause it simple, lite and sweet.
thank for this

File: src/utils/__test__/index.spec.tsx (hunk 17 above)
Comment (hoanglong-pyco) [reply src/utils/__test__/index.spec.tsx]: I was update this case for this in another PR.
Thank for comment.

File: src/utils/Store/Store.ts (hunk 15 above)
Comment (hoanglong-pyco) [reply src/utils/Store/Store.ts]: Oh! it is good idle, i will try for this
Thanks

File: src/utils/Store/Notifier.ts (hunk 14 above)
Comment (hoanglong-pyco) [reply src/utils/Store/Notifier.ts:34]: it's will throw error when remove a listener in other listener
ex: [listen1, listen2]
listen1 => remove listen2
run listen1 => array = [listen1] => missing listen2 but forEach is running.

but i trying compare performance of check undefined and slice to choose new solution.
Thank for ask

**Comment by hoanglong-pyco** [comment]
done
